### PR TITLE
Engagement cohorts

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes

*Please describe.*  
- relies on #4428. (Will repoint the base when it's merged)
- allow events to be used in defining cohorts
- allow for flexible time frames for event usage
- allow for counting number of event times (performed event X times in last Y days)
*If this affects the frontend, include screenshots.*  

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests
- [ ] Migrations are safe to run at scale (e.g. PostHog Cloud) – present proof if not obvious

### CODE REVIEW

File: ee/clickhouse/models/test/test_cohort.py [hunk 1]
```diff
@@ -321,6 +368,146 @@ def test_cohortpeople_action_basic(self):
...
+            groups=[{"action_id": action.pk, "start_date": datetime(2020, 1, 8, 12, 0, 1)}],
+            name="cohort1",
+        )
+        with freeze_time("2020-01-10"):
+            cohort1.calculate_people_ch()
+
+        results = sync_execute("SELECT person_id FROM cohortpeople")
+        self.assertEqual(len(results), 1)
+
+    def test_cohortpeople_action_count(self):
```
Comment (buwilliams) [inline ee/clickhouse/models/test/test_cohort.py:412]: Extract the mock code into a method with a good name.

Comment (buwilliams) [reply ee/clickhouse/models/test/test_cohort.py:412]: Overall, I feel this method could be broken up into smaller chunks.

File: ee/clickhouse/sql/cohort.py [hunk 2]
```diff
@@ -49,3 +49,33 @@
...
+GET_PERSON_ID_BY_ENTITY_COUNT_SQL = """
+SELECT person_id FROM events 
+INNER JOIN (
+    SELECT person_id,
+        distinct_id
+    FROM (
+            SELECT *
+            FROM person_distinct_id
+            JOIN (
+                    SELECT distinct_id,
```
Comment (buwilliams) [inline ee/clickhouse/sql/cohort.py:66]: We already have SQL to do this I believe.

**Review by buwilliams — CHANGES_REQUESTED** [review]

File: ee/clickhouse/models/test/test_cohort.py (hunk 1 above)
Comment (EDsCODE) [reply ee/clickhouse/models/test/test_cohort.py]: done

File: ee/clickhouse/sql/cohort.py (hunk 2 above)
Comment (EDsCODE) [reply ee/clickhouse/sql/cohort.py]: yup good point

**Review by buwilliams — APPROVED** [review]
lgtm

**Pull request merged**